VultisigApp/VultisigApp/Chains/cosmos/cosmos.swift

```diff
@@ -0,0 +1,246 @@
+            input.publicKey = pubKeyData
+            input.accountNumber = accountNumber
+            input.sequence = sequence
+            input.mode = .sync
+            
+            input.fee = CosmosFee.with {
+                $0.gas = self.gasLimit
+                $0.amounts = [CosmosAmount.with {
+                    $0.denom = self.denom
+                    $0.amount = String(gas)
+                }]
+            }
+            // memo has been set
+            // deposit message has been set
+            return try input.serializedData()
+        }
+        
+    
+    func getPreSignedInputData(keysignPayload: KeysignPayload) throws -> Data {
+        guard case .Cosmos(let accountNumber, let sequence , let gas, _, let ibc) = keysignPayload.chainSpecific else {
+            throw HelperError.runtimeError("getPreSignedInputData: fail to get account number and sequence")
+        }
+        guard let pubKeyData = Data(hexString: keysignPayload.coin.hexPublicKey) else {
+            throw HelperError.runtimeError("getPreSignedInputData: invalid hex public key")
+        }
+        let coin = self.coinType
+        
+        if keysignPayload.coin.isNativeToken
+            || keysignPayload.coin.contractAddress.lowercased().starts(with: "ibc/")
+            || keysignPayload.coin.contractAddress.lowercased().starts(with: "factory/")
+        {
+            
+            let input = CosmosSigningInput.with {
+                $0.publicKey = pubKeyData
+                $0.signingMode = .protobuf
+                $0.chainID = coin.chainId
+                $0.accountNumber = accountNumber
+                $0.sequence = sequence
+                $0.mode = .sync
+                if let memo = keysignPayload.memo {
+                    $0.memo = memo
+                }
+                $0.messages = [CosmosMessage.with {
+                    $0.sendCoinsMessage = CosmosMessage.Send.with{
+                        $0.fromAddress = keysignPayload.coin.address
+                        $0.amounts = [CosmosAmount.with {
+                            $0.denom = keysignPayload.coin.isNativeToken ? self.denom : keysignPayload.coin.contractAddress
+                            $0.amount = String(keysignPayload.toAmount)
+                        }]
+                        $0.toAddress = keysignPayload.toAddress
+                    }
+                }]
+                
+                $0.fee = CosmosFee.with {
+                    $0.gas = self.gasLimit
+                    $0.amounts = [CosmosAmount.with {
+                        $0.denom = self.denom
+                        $0.amount = String(gas)
+                    }]
+                }
+            }
+            
+            return try input.serializedData()
+            
+        }
+        
+        // IT seems this is only used when literally I want to move the coin from one blockchain to another
+        // IBCs are working as coins so we can simply use the send.
+//        else {
```

Comment:
clean up these

Reply:
Do we want to implement sending from one chain to another? I was thinking about add this feature to cosmos tomorrow. 

Cosmos1…. To osmo1… address

Reply:
https://github.com/vultisig/vultisig-ios/issues/1570 

Opened an issue and cleaned it up.